machine_learning/mfcc.py

```diff
@@ -0,0 +1,484 @@
+        freq_min,
+        freq_high,
+        mel_filter_num,
+        ftt_size,
+    )
+
+    filters = get_filters(filter_points, ftt_size)
+
+    # normalize filters
+    # taken from the librosa library
+    enorm = 2.0 / (mel_freqs[2 : mel_filter_num + 2] - mel_freqs[:mel_filter_num])
+    return filters * enorm[:, np.newaxis]
+
+
+def get_filters(filter_points: np.ndarray, ftt_size: int) -> np.ndarray:
+    """
+    Generate filters for audio processing.
+
+    Args:
+        filter_points: A list of filter points.
+        ftt_size: The size of the FFT.
+
+    Returns:
+        A matrix of filters.
+
+    Examples:
+    >>> get_filters(np.array([0, 20, 51, 95, 161, 256], dtype=int), 512).shape
+    (4, 257)
+    """
+    num_filters = len(filter_points) - 2
+    filters = np.zeros((num_filters, int(ftt_size / 2) + 1))
+
+    for n in range(num_filters):
+        start = filter_points[n]
+        mid = filter_points[n + 1]
+        end = filter_points[n + 2]
+
+        # Linearly increase values from 0 to 1
+        filters[n, start:mid] = np.linspace(0, 1, mid - start)
+
+        # Linearly decrease values from 1 to 0
+        filters[n, mid:end] = np.linspace(1, 0, end - mid)
+
+    return filters
+
+
+def get_filter_points(
+    sample_rate: int,
+    freq_min: int,
+    freq_high: int,
+    mel_filter_num: int = 10,
+    ftt_size: int = 1024,
+):
+    """
+    Calculate the filter points and frequencies for mel frequency filters.
+
+    Args:
+        sample_rate: The sample rate of the audio.
+        freq_min: The minimum frequency in Hertz.
+        freq_high: The maximum frequency in Hertz.
+        mel_filter_num: The number of mel filters (default is 10).
+        ftt_size: The size of the FFT (default is 1024).
+
+    Returns:
+        Filter points and corresponding frequencies.
+
+    Examples:
+    >>> get_filter_points(8000, 0, 4000, mel_filter_num=4, ftt_size=512)[0]
+    array([  0,  20,  51,  95, 161, 256])
+    """
+    # Convert minimum and maximum frequencies to mel scale
+    fmin_mel = freq_to_mel(freq_min)
+    fmax_mel = freq_to_mel(freq_high)
+
+    logging.info(f"MEL min: {fmin_mel}")
+    logging.info(f"MEL max: {fmax_mel}")
+
+    # Generate equally spaced mel frequencies
+    mels = np.linspace(fmin_mel, fmax_mel, num=mel_filter_num + 2)
+
+    # Convert mel frequencies back to Hertz
+    freqs = mel_to_freq(mels)
+
+    # Calculate filter points as integer values
+    filter_points = np.floor((ftt_size + 1) / sample_rate * freqs).astype(int)
+
+    return filter_points, freqs
+
+
+def discrete_cosine_transform(dct_filter_num: int, filter_num: int) -> np.ndarray:
+    """
+    Compute the Discrete Cosine Transform (DCT) basis matrix.
+
+    Args:
+        dct_filter_num: The number of DCT filters to generate.
+        filter_num: The number of the fbank filters.
+
+    Returns:
+        The DCT basis matrix.
+
+    Examples:
+    >>> round(discrete_cosine_transform(3, 5)[0][0], 5)
+    0.44721
```

Comment:
Lines 435 and 444 say that an array should be returned but this test says that a float should be returned.